cudamapper/src/main.cu

```diff
@@ -228,42 +229,49 @@ void process_one_device_batch(const IndexBatch& device_batch,
-
-                // Align overlaps
-                std::vector<std::string> cigars;
-                if (application_parameters.alignment_engines > 0)
+                try
                 {
-                    cigars.resize(overlaps.size());
-                    GW_NVTX_RANGE(profiler, "align_overlaps");
-                    align_overlaps(device_allocator,
-                                   overlaps,
-                                   *application_parameters.query_parser,
-                                   *application_parameters.target_parser,
-                                   application_parameters.alignment_engines,
-                                   cigars);
+                    // find anchors and overlaps
+                    auto matcher = Matcher::create_matcher(device_allocator,
+                                                           *query_index,
+                                                           *target_index,
+                                                           cuda_stream);
+
+                    std::vector<Overlap> overlaps;
+                    OverlapperTriggered overlapper(device_allocator,
+                                                   cuda_stream);
+                    overlapper.get_overlaps(overlaps,
+                                            matcher->anchors(),
+                                            application_parameters.min_residues,
+                                            application_parameters.min_overlap_len,
+                                            application_parameters.min_bases_per_residue,
+                                            application_parameters.min_overlap_fraction);
+
+                    // free up memory taken by matcher
+                    matcher.reset(nullptr);
+
+                    // Align overlaps
+                    std::vector<std::string> cigars;
+                    if (application_parameters.alignment_engines > 0)
+                    {
+                        cigars.resize(overlaps.size());
+                        GW_NVTX_RANGE(profiler, "align_overlaps");
+                        align_overlaps(device_allocator,
+                                       overlaps,
+                                       *application_parameters.query_parser,
+                                       *application_parameters.target_parser,
+                                       application_parameters.alignment_engines,
+                                       cigars);
+                    }
+
+                    // pass overlaps and cigars to writer thread
+                    overlaps_and_cigars_to_process.add_new_element({std::move(overlaps), std::move(cigars)});
+                }
+                catch (device_memory_allocation_exception& oom_exception)
+                {
+                    // if the application ran out of memory skip this pair of indices
+                    ++(number_of_skipped_pairs_of_indices);
```

Comment:
Do we need this whole section in the `try` clause? I thought the oom_exception was only thrown by create_matcher. My only concern is that this could catch other oom_exceptions beyond the one we intend to catch with this PR.

Reply:
Actually that was my intention. Device OOM can happen in matcher, overlapper and aligner and we would like to catch them all. It's actually all the same case: running out of memory while processing a pair of indices.
Does that make sense?

Reply:
Yep, that makes sense and is cleaner than individual try-catch clauses for each call.